### PR TITLE
[5.1] Added feature for issue #11319

### PR DESCRIPTION
**Pull request by lianguan**

### CODE REVIEW

**Comment by phecho** [comment]
usage:

```html
<a class="btn btn-default" href="{{ back_url('groups.group_show', ['group' => $group->path]) }}">{{ trans('forms.cancel') }}</a>
```

File: src/Illuminate/Foundation/helpers.php [hunk 1]
```diff
@@ -115,6 +115,26 @@ function back($status = 302, $headers = [])
     }
 }
 
+if (! function_exists('back_url')) {
+    /**
+     * Create a new redirect url to the previous location.
+     *
+     * @param string $route
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: cs

Comment (phecho) [inline src/Illuminate/Foundation/helpers.php]: @GrahamCampbell 

thanks, done.

File: src/Illuminate/Foundation/helpers.php [hunk 2]
```diff
@@ -115,6 +115,26 @@ function back($status = 302, $headers = [])
     }
 }
 
+if (! function_exists('back_url')) {
+    /**
+     * Create a new redirect url to the previous location.
+     *
+     * @param  string $route
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: cs

Comment (phecho) [inline src/Illuminate/Foundation/helpers.php]: @GrahamCampbell 
is ok now?

File: src/Illuminate/Foundation/helpers.php [hunk 3]
```diff
@@ -115,6 +115,26 @@ function back($status = 302, $headers = [])
     }
 }
 
+if (! function_exists('back_url')) {
+    /**
+     * Create a new redirect url to the previous location.
+     *
+     * @param  string  $route
+     * @param  array   $parameters
+     * @param  int     $status
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: don't align please

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: just 2 spaces each side of the annotation

Comment (phecho) [inline src/Illuminate/Foundation/helpers.php]: It is confused. I saw codes below in helpers.php.
```php
f (! function_exists('cookie')) {
    /**
     * Create a new cookie instance.
     *
     * @param  string  $name
     * @param  string  $value
     * @param  int     $minutes
     * @param  string  $path
     * @param  string  $domain
     * @param  bool    $secure
     * @param  bool    $httpOnly
     * @return \Symfony\Component\HttpFoundation\Cookie
     */
    function cookie($name = null, $value = null, $minutes = 0, $path = null, $domain = null, $secure = false, $httpOnly = true)
    {
        $cookie = app(CookieFactory::class);

        if (is_null($name)) {
            return $cookie;
        }

        return $cookie->make($name, $value, $minutes, $path, $domain, $secure, $httpOnly);
    }
}
```

**Comment by phecho** [comment]
@GrahamCampbell 
I read the official docs on laravel.com, you are right.

I fixed the CS typos in this file.

**Comment by GrahamCampbell** [comment]
Yeh, I wrote those rules. ;)

Please don't modify the other functions in the file though. These rules only apply to new code, not existing code to avoid diffs changing whitespace everywherw.

**Comment by phecho** [comment]
@GrahamCampbell 
OK, I reset the commit.

File: src/Illuminate/Foundation/helpers.php [hunk 4]
```diff
@@ -115,6 +115,26 @@ function back($status = 302, $headers = [])
...
+    /**
+     * Create a new redirect url to the previous location.
+     *
+     * @param  string  $route
+     * @param  array  $parameters
+     * @param  int  $status
+     * @param  array  $headers
+     * @return string
+     */
+    function back_url($route, $parameters = [], $status = 302, $headers = [])
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: we need a default param of `null` on `$route`

File: src/Illuminate/Foundation/helpers.php [hunk 5]
```diff
@@ -115,6 +115,26 @@ function back($status = 302, $headers = [])
...
+    {
+        if ($route !== null && app('url')->previous() == app('url')->full()) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: please use `! is_null`

File: src/Illuminate/Foundation/helpers.php [hunk 6]
```diff
@@ -115,6 +115,26 @@ function back($status = 302, $headers = [])
     }
 }
 
+if (! function_exists('back_url')) {
+    /**
+     * Create a new redirect url to the previous location.
+     *
+     * @param  string  $route
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: `string|null`

File: src/Illuminate/Foundation/helpers.php (hunk 5 above)
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: please only call `app('url') once

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/helpers.php]: maybe just assign `$url = app('url')` at the top of the function?

**Comment by phecho** [comment]
@GrahamCampbell 
I think  all done.
Thank you very much.

**Comment by rkgrep** [comment]
I don't like that it is impossible to use `action` or just other url.
Why don't make this usage
```php
back_url(route('my.route'));
back_url(action('My@action));
back_url('some_other_url');
```

**Comment by rkgrep** [comment]
An `previous` helper sounds much better IMO.

**Comment by phecho** [comment]
@rkgrep 
how about code bellow?
```php
if (!function_exists('previous')) {
    /**
     * Create a new redirect url to the previous location.
     *
     * @param string|null $default
     *
     * @return string
     */
    function previous($default = null)
    {
        $url = app('url');

        if (!is_null($default) && $url->previous() == $url->full()) {
            return $default;
        }

        return $url->previous();
    }
}
```

**Comment by taylorotwell** [comment]
Just use URL::previous().

**Pull request closed without merging**